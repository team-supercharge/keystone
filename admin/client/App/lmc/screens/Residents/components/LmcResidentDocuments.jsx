import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { ActionCreators } from '../../../actions/actions'
import { BlankState } from '../../../../elemental'
import Selectors from '../../../selectors'
import LmcDocumentItem from './LmcDocumentItem.jsx'
import LmcCreateButton from '../../../components/LmcCreateButton.jsx'

export class LmcResidentDocuments extends Component {
    state = {
        documentsFetchInterval: null
    }

    componentDidMount () {
        const { fetchDocuments } = this.props
        this.setState({ 
            documentsFetchInterval: setInterval(fetchDocuments, 600000)
        })
        fetchDocuments()
    }

    componentWillUnmount() {
        clearInterval(this.state.documentsFetchInterval)
    }

    listDocumentsByCategory = category => {
        const { documents, deleteDocument } = this.props
        return (
            documents[category].map((document, i) => (
                <LmcDocumentItem
                    key={i}
                    data={document}
                    onDelete={deleteDocument}
                />
            ))
        )
    }

    renderDocuments = () => {
        return (
            Object.keys(this.props.documents).map(categoryName => (
                <div key={categoryName}>
                    <h2 style={styles.categoryName}>
                        {categoryName}
                    </h2>
                    <div className='lmc-theme-gradient' style={styles.divider} />
                    <ul style={styles.list}>
                        { this.listDocumentsByCategory(categoryName) }
                    </ul>
                </div>
            ))
        )
    }

    render () {
        const { documents, fetchDocuments, selectedResident } = this.props
        const hasDocuments = !!Object.keys(documents).length

        return (
            <div>
                <LmcCreateButton
                    buttonText='Document'
                    listId='Document'
                    title='Add a Document'
                    onCreate={fetchDocuments}
                    prefillPath='resident'
                    prefillValue={selectedResident}
                    style={styles.addButton}
                />
                { hasDocuments ? (
                    <div>
                        { this.renderDocuments() }
                    </div>
                ) : (
                    <BlankState
                        heading={NO_DOCUMENTS_MESSAGE}
                        style={styles.noDocumentsMessage}
                    />
                ) }
            </div>
        )
    }
}

const NO_DOCUMENTS_MESSAGE = "You haven't added any documents for this resident"

const styles = {
    addButton: {
        float: 'right',
        width: 200,
    },
    categoryName: {
        marginBottom: '0.3em',
        fontWeight: 300,
    },
    divider: {
        height: 2,
        marginBottom: 22,
        width: '100%',
    },
    documentContainer: {
        display: 'inline-block',
    },
    list: {
        listStyle: 'none',
        listStyleType: 'none',
        padding: 0,
    },
    noDocumentsMessage: {
        position: 'relative',
        top: 50,
    },
}

LmcResidentDocuments.propTypes = {
    documents: PropTypes.object,
    fetchDocuments: PropTypes.func.isRequired,
    deleteDocument: PropTypes.func.isRequired,
    selectedResident: PropTypes.string,
}

const mapStateToProps = (state) => {
    return {
        documents: Selectors.getSelectedResidentDocuments(state),
        selectedResident: state.residents.selectedResident,
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        fetchDocuments: () => dispatch(ActionCreators.loadList('documents')),
        deleteDocument: (id) => dispatch(ActionCreators.deleteDocument(id)),
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(LmcResidentDocuments)