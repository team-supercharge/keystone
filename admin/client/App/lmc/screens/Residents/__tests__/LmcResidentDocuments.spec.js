jest.mock('../../../../shared/CreateForm')

import React from 'react'
import { shallow } from 'enzyme'
import { LmcResidentDocuments } from '../components/LmcResidentDocuments.jsx'

describe('LmcResidentDocuments', () => {
    let wrapper
    let documents
    const deleteDocument = jest.fn()
    const fetchDocuments = jest.fn()

    beforeEach(() => {
        documents = { 'Going to Bed and Sleeping': [{ name: 'TestDocument' }]}
        wrapper = shallow(
            <LmcResidentDocuments 
                documents={documents}
                fetchDocuments={fetchDocuments}
                deleteDocument={deleteDocument}
            />
        )
    })

    it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot()
    })

    it('should display a message if no documents are present', () => {
        const emptyWrapper = shallow(
            <LmcResidentDocuments
                documents={{}}
                fetchDocuments={fetchDocuments}
                deleteDocument={deleteDocument}
            />
        )
        const message = emptyWrapper.find('BlankState')
        expect(message.props().heading).toEqual("You haven't added any documents for this resident")
    })

    it('should display an LmcDocumentItem component with the correct data', () => {
        const documentItem = wrapper.find('LmcDocumentItem').first()
        expect(documentItem.props().data).toEqual(documents['Going to Bed and Sleeping'][0])

        documentItem.props().onDelete()
        expect(deleteDocument).toBeCalledTimes(1)
    })

    it('renders a LmcCreateButton with the correct props', () => {
        const button = wrapper.find('LmcCreateButton')

        expect(button.props().listId).toEqual('Document')
        expect(button.props().title).toEqual('Add a Document')
        expect(button.props().onCreate).toEqual(fetchDocuments)
    })

    afterEach(() => {
        fetchDocuments.mockClear()
        deleteDocument.mockClear()
    })
})