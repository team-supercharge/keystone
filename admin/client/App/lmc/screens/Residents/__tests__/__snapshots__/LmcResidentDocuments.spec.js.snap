// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LmcResidentDocuments renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <LmcCreateButton
      buttonText="Document"
      listId="Document"
      onCreate={[MockFunction]}
      prefillPath="resident"
      style={
        Object {
          "float": "right",
          "width": 200,
        }
      }
      title="Add a Document"
    />
    <div>
      <div>
        <h2
          style={
            Object {
              "fontWeight": 300,
              "marginBottom": "0.3em",
            }
          }
        >
          Going to Bed and Sleeping
        </h2>
        <div
          className="lmc-theme-gradient"
          style={
            Object {
              "height": 2,
              "marginBottom": 22,
              "width": "100%",
            }
          }
        />
        <ul
          style={
            Object {
              "listStyle": "none",
              "listStyleType": "none",
              "padding": 0,
            }
          }
        >
          <LmcDocumentItem
            data={
              Object {
                "name": "TestDocument",
              }
            }
            onDelete={[MockFunction]}
          />
        </ul>
      </div>
    </div>
  </div>,
  "nodes": Array [
    <div>
      <LmcCreateButton
        buttonText="Document"
        listId="Document"
        onCreate={[MockFunction]}
        prefillPath="resident"
        style={
          Object {
            "float": "right",
            "width": 200,
          }
        }
        title="Add a Document"
      />
      <div>
        <div>
          <h2
            style={
              Object {
                "fontWeight": 300,
                "marginBottom": "0.3em",
              }
            }
          >
            Going to Bed and Sleeping
          </h2>
          <div
            className="lmc-theme-gradient"
            style={
              Object {
                "height": 2,
                "marginBottom": 22,
                "width": "100%",
              }
            }
          />
          <ul
            style={
              Object {
                "listStyle": "none",
                "listStyleType": "none",
                "padding": 0,
              }
            }
          >
            <LmcDocumentItem
              data={
                Object {
                  "name": "TestDocument",
                }
              }
              onDelete={[MockFunction]}
            />
          </ul>
        </div>
      </div>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <LmcResidentDocuments
        deleteDocument={[MockFunction]}
        documents={
          Object {
            "Going to Bed and Sleeping": Array [
              Object {
                "name": "TestDocument",
              },
            ],
          }
        }
        fetchDocuments={[MockFunction]}
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": LmcResidentDocuments {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "listDocumentsByCategory": [Function],
        "props": Object {
          "deleteDocument": [MockFunction],
          "documents": Object {
            "Going to Bed and Sleeping": Array [
              Object {
                "name": "TestDocument",
              },
            ],
          },
          "fetchDocuments": [MockFunction],
        },
        "refs": Object {},
        "renderDocuments": [Function],
        "state": Object {
          "documentsFetchInterval": null,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          <LmcCreateButton
            buttonText="Document"
            listId="Document"
            onCreate={[MockFunction]}
            prefillPath="resident"
            style={
              Object {
                "float": "right",
                "width": 200,
              }
            }
            title="Add a Document"
          />
          <div>
            <div>
              <h2
                style={
                  Object {
                    "fontWeight": 300,
                    "marginBottom": "0.3em",
                  }
                }
              >
                Going to Bed and Sleeping
              </h2>
              <div
                className="lmc-theme-gradient"
                style={
                  Object {
                    "height": 2,
                    "marginBottom": 22,
                    "width": "100%",
                  }
                }
              />
              <ul
                style={
                  Object {
                    "listStyle": "none",
                    "listStyleType": "none",
                    "padding": 0,
                  }
                }
              >
                <LmcDocumentItem
                  data={
                    Object {
                      "name": "TestDocument",
                    }
                  }
                  onDelete={[MockFunction]}
                />
              </ul>
            </div>
          </div>
        </div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          <LmcCreateButton
            buttonText="Document"
            listId="Document"
            onCreate={[MockFunction]}
            prefillPath="resident"
            style={
              Object {
                "float": "right",
                "width": 200,
              }
            }
            title="Add a Document"
          />
          <div>
            <div>
              <h2
                style={
                  Object {
                    "fontWeight": 300,
                    "marginBottom": "0.3em",
                  }
                }
              >
                Going to Bed and Sleeping
              </h2>
              <div
                className="lmc-theme-gradient"
                style={
                  Object {
                    "height": 2,
                    "marginBottom": 22,
                    "width": "100%",
                  }
                }
              />
              <ul
                style={
                  Object {
                    "listStyle": "none",
                    "listStyleType": "none",
                    "padding": 0,
                  }
                }
              >
                <LmcDocumentItem
                  data={
                    Object {
                      "name": "TestDocument",
                    }
                  }
                  onDelete={[MockFunction]}
                />
              </ul>
            </div>
          </div>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <LmcResidentDocuments
    deleteDocument={[MockFunction]}
    documents={
      Object {
        "Going to Bed and Sleeping": Array [
          Object {
            "name": "TestDocument",
          },
        ],
      }
    }
    fetchDocuments={[MockFunction]}
  />,
}
`;
